src/mqtt_client.rs

```diff
@@ -239,12 +250,16 @@ where
         let mut buffer = self.transmit_buffer.borrow_mut();
         let packet = serialize::publish_message(&mut buffer, topic, data, properties)
             .map_err(|e| Error::Protocol(e))?;
-        self.write(packet)
+        self.write(&mut self.state.borrow_mut(), packet)
     }
 
     fn socket_is_connected(&self) -> Result<bool, N::Error> {
         let mut socket_ref = self.socket.borrow_mut();
-        let socket = socket_ref.take().unwrap();
+        let socket = socket_ref.take();
+        if socket.is_none() {
```

Comment:
This should be treated as an error condition, as it indicates we have a bug in the library.

Reply:
I don't think so, as `open` is allowed to fail. `shutdown_socket` thus just pulls out the socket entirely without replacing it.